### PR TITLE
[BEAM-11929] Rely on py3.6+ dictionary ordering in beam.Row

### PR DESCRIPTION
**Pull request by TheNeuralBit**
Previously fields were sorted by name, which lead to a mismatch between the inferred schema and some interactions with the Row object (e.g. iterating), leading to bugs like BEAM-11929. With this change we just rely on the consistent dictionary ordering provided in python 3.6+ to provide a consistent field order.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by TheNeuralBit** [comment]
R: @udim 
CC: @robertwb

**Comment by udim** [comment]
There's a failing test: https://ci-beam.apache.org/job/beam_PreCommit_Python_Commit/17554/testReport/junit/apache_beam.dataframe.io_test/IOTest/test_read_write_parquet/

File: sdks/python/apache_beam/pvalue.py [hunk 1]
```diff
@@ -666,15 +666,14 @@ def as_dict(self):
...
       yield value
 
   def __repr__(self):
-    return 'Row(%s)' % ', '.join(
-        '%s=%r' % kv for kv in sorted(self.__dict__.items()))
+    return 'Row(%s)' % ', '.join('%s=%r' % kv for kv in self.__dict__.items())
 
   def __hash__(self):
-    return hash(type(sorted(self.__dict__.items())))
+    return hash(type(self.__dict__.items()))
```
Comment (udim) [inline sdks/python/apache_beam/pvalue.py:676]: I tried running this interactively but got the same hash:
```
>>> hash(type({}.items()))
5913863196444
>>> hash(type({1:2}.items()))
5913863196444
```

Is it working as intended?

I guess this is technically okay according to the docs, but probably not what you wanted: `The only required property is that objects which compare equal have the same hash value`.
[ref](https://docs.python.org/3/reference/datamodel.html#object.__hash__)

File: sdks/python/apache_beam/pvalue.py [hunk 2]
```diff
@@ -666,15 +666,14 @@ def as_dict(self):
...
 
   def __eq__(self, other):
     return type(self) == type(other) and self.__dict__ == other.__dict__
```
Comment (udim) [inline sdks/python/apache_beam/pvalue.py:679]: Comparing dicts ignores order. Is that that's intentional?
```
>>> {1:2, 3:4} == {1:2, 3:4}
True
>>> {1:2, 3:4} == {3:4, 1:2}
True
```

**Review by udim — COMMENTED** [review]
This change should be mentioned in CHANGES.md if it breaks with existing behavior. For instance, if existing uses rely on fields being sorted by name.

File: sdks/python/apache_beam/pvalue.py (hunk 1 above)
Comment (TheNeuralBit) [reply sdks/python/apache_beam/pvalue.py:676]: I didn't write the original `__hash__` implementation that uses `type()`. You're right that it seems like an odd choice though. It seems likely this was a copy-paste error or a typo.

File: sdks/python/apache_beam/pvalue.py (hunk 2 above)
Comment (TheNeuralBit) [reply sdks/python/apache_beam/pvalue.py:679]: Hm that's consistent with the current implementation, but IMO it makes sense to make this sensitive to field order. That will be consistent with Java, where two Rows with different schemas are considered unequal.

**Comment by TheNeuralBit** [comment]
Run Python PreCommit

**Comment by TheNeuralBit** [comment]
> There's a failing test: https://ci-beam.apache.org/job/beam_PreCommit_Python_Commit/17554/testReport/junit/apache_beam.dataframe.io_test/IOTest/test_read_write_parquet/

That's a weird failure. It doesn't seem related to this change and I can't repro it locally. But it's also not a flake that shows up on Python PreCommit Cron. Re-running PreCommit on  unmodified code to see if it's reproducible.

File: sdks/python/apache_beam/pvalue.py (hunk 2 above)
Comment (TheNeuralBit) [reply sdks/python/apache_beam/pvalue.py]: Done.

File: sdks/python/apache_beam/pvalue.py (hunk 1 above)
Comment (TheNeuralBit) [reply sdks/python/apache_beam/pvalue.py]: Ok, went ahead and pushed a commit to fix this.

**Comment by TheNeuralBit** [comment]
Ok, I think I've addressed all the comments, PTAL @udim

**Comment by TheNeuralBit** [comment]
ping @ehudm do you have time to review this?

**Review by udim — APPROVED** [review]
Thanks, and sorry for the delay!

**Pull request merged**